pkg/api/tag.go

```diff
@@ -0,0 +1,203 @@
+	if err != nil {
+		s.Logger.Debugf("get tag: parse id  %s: %v", idStr, err)
+		s.Logger.Error("get tag: parse id")
+		jsonhttp.BadRequest(w, "invalid id")
+		return
+	}
+
+	tag, err := s.Tags.Get(uint32(id))
+	if err != nil {
+		if errors.Is(err, tags.ErrNotFound) {
+			s.Logger.Debugf("get tag: tag not present: %v, id %s", err, idStr)
+			s.Logger.Error("get tag: tag not present")
+			jsonhttp.NotFound(w, "tag not present")
+			return
+		}
+		s.Logger.Debugf("get tag: tag %v: %v", idStr, err)
+		s.Logger.Errorf("get tag: %v", idStr)
+		jsonhttp.InternalServerError(w, "cannot get tag")
+		return
+	}
+
+	w.Header().Set("Cache-Control", "no-cache, private, max-age=0")
+	jsonhttp.OK(w, newTagResponse(tag))
+}
+
+func (s *server) deleteTag(w http.ResponseWriter, r *http.Request) {
+	idStr := mux.Vars(r)["id"]
+
+	id, err := strconv.Atoi(idStr)
+	if err != nil {
+		s.Logger.Debugf("delete tag: parse id  %s: %v", idStr, err)
+		s.Logger.Error("delete tag: parse id")
+		jsonhttp.BadRequest(w, "invalid id")
+		return
+	}
+
+	tag, err := s.Tags.Get(uint32(id))
+	if err != nil {
+		if errors.Is(err, tags.ErrNotFound) {
+			s.Logger.Debugf("delete tag: tag not present: %v, id %s", err, idStr)
+			s.Logger.Error("delete tag: tag not present")
+			jsonhttp.NotFound(w, "tag not present")
+			return
+		}
+		s.Logger.Debugf("delete tag: tag %v: %v", idStr, err)
+		s.Logger.Errorf("delete tag: %v", idStr)
+		jsonhttp.InternalServerError(w, "cannot get tag")
+		return
+	}
+
+	s.Tags.Delete(tag.Uid)
+	jsonhttp.NoContent(w)
+}
+
+func (s *server) doneSplit(w http.ResponseWriter, r *http.Request) {
+	idStr := mux.Vars(r)["id"]
+
+	id, err := strconv.Atoi(idStr)
+	if err != nil {
+		s.Logger.Debugf("done split tag: parse id  %s: %v", idStr, err)
+		s.Logger.Error("done split tag: parse id")
+		jsonhttp.BadRequest(w, "invalid id")
+		return
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		if jsonhttp.HandleBodyReadError(err, w) {
+			return
+		}
+		s.Logger.Debugf("done split tag: read request body error: %v", err)
+		s.Logger.Error("done split tag: read request body error")
+		jsonhttp.InternalServerError(w, "cannot read request")
+		return
+	}
+
+	tagr := tagRequest{}
+	if len(body) > 0 {
+		err = json.Unmarshal(body, &tagr)
+		if err != nil {
+			s.Logger.Debugf("done split tag: unmarshal tag name error: %v", err)
+			s.Logger.Errorf("done split tag: unmarshal tag name error")
+			jsonhttp.InternalServerError(w, "error unmarshaling metadata")
+			return
+		}
+	}
+
+	tag, err := s.Tags.Get(uint32(id))
+	if err != nil {
+		if errors.Is(err, tags.ErrNotFound) {
+			s.Logger.Debugf("done split: tag not present: %v, id %s", err, idStr)
+			s.Logger.Error("done split: tag not present")
+			jsonhttp.NotFound(w, "tag not present")
+			return
+		}
+		s.Logger.Debugf("done split: tag %v: %v", idStr, err)
+		s.Logger.Errorf("done split: %v", idStr)
+		jsonhttp.InternalServerError(w, "cannot get tag")
+		return
+	}
+
+	tag.DoneSplit(tagr.Address)
```

Comment:
what address are we using here??

Reply:
the address that the api client has called the endpoint with.
two scenarios:
- i am an API client and I have multiple uploads that have different root hashes but there is not one canonical root hash that represents all of these uploads, then I call this api with an empty body. this calls the `DoneSplit` and sets `Split = Total` but the root hash is empty (because there is not a root hash)
- i am an API client and I have one single upload (or multiple) that can be represented by one root hash. I call this api endpoint with the JSON request and set the root hash of the tag to the one which represents my uploads.

Reply:
> the address that the api client has called the endpoint with.
> two scenarios:
> 
>     * i am an API client and I have multiple uploads that have different root hashes but there is not one canonical root hash that represents all of these uploads, then I call this api with an empty body. this calls the `DoneSplit` and sets `Split = Total` but the root hash is empty (because there is not a root hash)

i'm a bit confused about this... should we be allowing `DoneSplit` calls with an empty address? 

because currently we are not